src/Processors/Transforms/WindowTransform.cpp

```diff
@@ -2200,6 +2203,128 @@ struct WindowFunctionNthValue final : public WindowFunction
+                            argument_types[ARGUMENT_TIMESTAMP]->getName());
+        }
+
+        if (typeid_cast<const DataTypeDateTime64 *>(argument_types[ARGUMENT_TIMESTAMP].get()))
+        {
+            const auto & datetime64_type = assert_cast<const DataTypeDateTime64 &>(*argument_types[ARGUMENT_TIMESTAMP]);
+            ts_scale_multiplier = DecimalUtils::scaleMultiplier<DateTime64>(datetime64_type.getScale());
+        }
+
+        if (argument_types.size() == 3)
+        {
+            const DataTypeInterval * interval_datatype = checkAndGetDataType<DataTypeInterval>(argument_types[ARGUMENT_INTERVAL].get());
+            if (!interval_datatype)
+            {
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS,
+                    "Argument {} must be an INTERVAL, '{}' given",
+                    ARGUMENT_INTERVAL,
+                    argument_types[ARGUMENT_INTERVAL]->getName());
+            }
+            if (!interval_datatype->getKind().isFixedLength())
+            {
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "The INTERVAL must be a week or shorter, '{}' given",
+                    argument_types[ARGUMENT_INTERVAL]->getName());
+            }
+            interval_length = interval_datatype->getKind().toAvgSeconds();
+            interval_specified = true;
+        }
+    }
+
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeFloat64>(); }
+
+    bool allocatesMemoryInArena() const override { return false; }
+
+    void windowInsertResultInto(const WindowTransform * transform,
+                                size_t function_index) override
+    {
+        const auto & current_block = transform->blockAt(transform->current_row);
+        const auto & workspace = transform->workspaces[function_index];
+        auto & state = getState(workspace);
+
+        auto interval_duration = interval_specified ? interval_length *
+            (*current_block.input_columns[workspace.argument_column_indices[ARGUMENT_INTERVAL]]).getFloat64(0) : 1;
+
+        Float64 curr_metric = WindowFunctionHelpers::getValue<Float64>(transform, function_index, ARGUMENT_METRIC, transform->current_row);
+        Float64 metric_diff = curr_metric - state.previous_metric;
+        Float64 result;
+
+        if (ts_scale_multiplier)
+        {
+            const auto & column = transform->blockAt(transform->current_row.block).input_columns[workspace.argument_column_indices[ARGUMENT_TIMESTAMP]];
+            const auto & curr_timestamp = checkAndGetColumn<DataTypeDateTime64::ColumnType>(column.get())->getInt(transform->current_row.row);
+
+            Float64 time_elapsed = curr_timestamp - state.previous_timestamp;
+            result = (time_elapsed > 0) ? (metric_diff * ts_scale_multiplier / time_elapsed  * interval_duration) : 0;
+            state.previous_timestamp = curr_timestamp;
```

Comment:
Could pull l. 2306 / 2313 out, similar to l. 2315.

Reply:
No, we cannot -- `curr_timestamp` is initialized differently in the branches